src/engine/sparqlExpressions/StringExpressions.cpp

```diff
@@ -217,13 +239,14 @@ class SubstrImpl {
       return static_cast<size_t>(n);
     };
 
-    return toLiteral(
-        ad_utility::getUTF8Substring(str, clamp(startInt), clamp(lengthInt)));
+    return toLiteralWithDescriptor(
+        ad_utility::getUTF8Substring(str, clamp(startInt), clamp(lengthInt)),
+        descriptor);
   }
 };
 
-using SubstrExpression =
-    StringExpressionImpl<3, SubstrImpl, NumericValueGetter, NumericValueGetter>;
+using SubstrExpression = NARY<3, FV<SubstrImpl, LiteralOrIriValueGetter,
+                                    NumericValueGetter, NumericValueGetter>>;
```

Comment:
To make the case `SUBSTR(STR(?something))` more efficient, you can implement something like the `StringExpressionImpl`, which does exactly the same thing, but uses your two value getters (the one is still missing), depending on whether the child is a string or not.

Reply:
I tried to make a SubStrExpressionImpl like StringExpressionImpl, but then I get a lot of code duplications. Then I thought that SubStrExpressionImpl could inherit most of the code from StringExpressionImpl, but that doesn't fit conceptually, because here( using ExpressionWithStr = NARY<N,FV<Function, StringValueGetter, AdditionalNonStringValueGetters...>>; ) StringValueGetter is used directly. What is the best way to do it?